mqtt-simulator/simulator.py

```diff
@@ -11,26 +11,33 @@ def __init__(self, settings_file):
     def load_settings(self, settings_file):
         with open(settings_file) as json_file:
             config = json.load(json_file)
-            self.broker_url = config['BROKER_URL']
-            self.broker_port = config['BROKER_PORT']
+            self.broker_url = config.get('BROKER_URL', 'localhost')
+            self.broker_port = config.get('BROKER_PORT', 1883)
+            broker_time_interval = config.get('TIME_INTERVAL', 10)
+            broker_protocol = config.get('PROTOCOL_VERSION', 4) # mqtt.MQTTv311
+            broker_clean = config.get('CLEAN_SESSION', True)
+            broker_qos = config.get('QOS', 2)
+            broker_retained = config.get('RETAIN', False)
             # read each configured topic
             for topic in config['TOPICS']:
                 topic_data = topic['DATA']
-                topic_time_interval = topic['TIME_INTERVAL']
+                topic_time_interval = topic.get('TIME_INTERVAL', broker_time_interval)
+                topic_qos = topic.get('QOS', broker_qos)
+                topic_retained = topic.get('RETAIN', broker_retained)
```

Comment:
Any reason for removing the clean property from topic?
Since retain and clean seem related, I think if one exists it makes sense for both to exist.
Let me know what you think, if it makes sense please add `topic_clean` again (and update the docs).
Also, could you rename `broker_retained` and `topic_retained` to `broker_retain` and `topic_retain`?

As for topic_protocol, it's up to you, I don't know if it's necessary to exist here. If you think it doesn't make sense, you don't need to add it again.

Reply:
I updated the naming, thank you for catching that.

I removed the topic level clean anv version after reading the docs for the mqtt [client](https://pypi.org/project/paho-mqtt/#constructor-reinitialise) which manages those vs the the [publish](https://pypi.org/project/paho-mqtt/#publishing) function which takes the qos and retain flags. 

I think if you want to separate the files it will makes sense to decide of the simulator is designed to manage multiple unique broker configs or one broker with multilpe `publish` configs. For instance would one config manage connections on two URLs and Versions. IMO that would just be two different configs and two executions of the script. 

I am happy to make any changes you prefer, but we can also address this in a later PR aswell.

Reply:
Got it. I'll read about this later and maybe make further changes after this PR.

Reply:
@vordimous one last thing, you forgot to rename from `topic_retained` to `topic_retain` when instantiating `TopicAuto`.

Reply:
I fixed it so I can merge now 👍